### PR TITLE
Fix permissions breaking after rails reloads

### PR DESCRIPTION
**Pull request by xjunior**

### CODE REVIEW

File: lib/consent/railtie.rb [hunk 1]
```diff
@@ -8,7 +8,10 @@ class Railtie < Rails::Railtie
 
     config.to_prepare do
       Consent.subjects.clear
-      Consent.load_subjects! Rails.application.config.consent.paths
+      Rails.application
```
Comment (gmfvpereira) [inline lib/consent/railtie.rb]: How did you know that this was the problem?

Comment (xjunior) [reply lib/consent/railtie.rb]: @gmfvpereira `subjects.clear`would clear all the subjects `Consent` knows, while `load_subjects` would only `require` the files back. The problem is that `require` only loads the file once so they weren't being reloaded.

What puzzled me is that `to_prepare` was being called only on certain circumstances, and I would expect it to be called when a model changes, for instance.

Comment (todtb) [reply lib/consent/railtie.rb]: So if 

> to_prepare will run upon every request in development

How important is it to load these fresh each time in dev?

Comment (xjunior) [reply lib/consent/railtie.rb]: I guess it decides when to reload somehow, I didn't dig much as this is truly the main glue of the rails framework (constant/files autoloading). But we need this hot spot to be able to reload the permission files in `development` to allow a better development environment.

It's important to note that leaving it to the `autoload_paths` wouldn't work for two reasons:

1. We'd still need to prime load the permission files so they're available through `Consent.subjects`;
2. The lazy loading of them is attached to a constant autoloading, which isn't the case as each permission file does not define a module/class/constant, but represent a permission definition through a DSL;
3. Even if it defined a constant that constant would have to be mentioned elsewhere, which for this purpose would impose a bureaucratic step to load it (as the permission definition is not needed externally by any purpose but internal usage within consent itself);

There are other reasons as well as two using constant autoloading wasn't an option here and therefore the `railtie` hotspot was used to provide autoloading, but the above are the more important ones.

Comment (todtb) [reply lib/consent/railtie.rb]: There are many ways we can go about it for sure. My big concern is this:

> But we need this hot spot to be able to reload the permission files in development to allow a better development environment.

Restarting the server on permissions changes gives a far better development environment than restarting the server on each request. Therefore, I don't see having these reload as a priority. If we can do this while at the same time maintaining ease of development throughout the app, that's ok, the only concern is the marginal performance hit we take on every request in development environment as these reload.

Comment (xjunior) [reply lib/consent/railtie.rb]: I see this type of reloading is the norm in rails in `development`.  The problem here is that even if we set up a file watcher to know which files to reload, we'd also have to know which subjects to remove from the current definitions loaded to Consent.

I must also add that `to_prepare` is only called once in `production`.

Comment (xjunior) [reply lib/consent/railtie.rb]: @todtb I've researched how stuff are implemented internally in rails and found a plugin API for updaters. It's unfortunately very badly documented, but in the end, I was able to introduce an updater that will only reload permissions if they change (but always reloads all if one changes):

https://github.com/powerhome/consent/pull/6/commits/e354fcd1f9a637aa10ad35b4847e51c9e8bffc37

Comment (todtb) [reply lib/consent/railtie.rb]: Nice find! That's pretty cool, never heard of it either, but looks handy. From what I can tell this satisfies both permission changes and app changes in dev.

**Pull request merged**